src/emqx_auth_ldap.erl

```diff
@@ -62,14 +64,67 @@ check(ClientInfo = #{username := Username, password := Password}, AuthResult,
             {stop, AuthResult#{auth_result => ResultCode, anonymous => false}}
     end.
 
-lookup_user(Username, #{username_attr := UidAttr,
-                        match_objectclass := ObjectClass,
-                        device_dn := DeviceDn}) ->
-    Filter = eldap2:equalityMatch("objectClass", ObjectClass),
-    case search(lists:concat([UidAttr,"=", binary_to_list(Username), ",", DeviceDn]), Filter) of
-        {error, noSuchObject} ->
+lookup_user(Username, Password, #{username_attr := UidAttr,
+                                  match_objectclass := ObjectClass,
+                                  device_dn := DeviceDn,
+                                  bind_as_user := BindAsUserRequired,
+                                  custom_base_dn := CustomBaseDN} = Config) ->
+
+    Filters = maps:get(filters, Config, []),
+
+    ReplaceRules = [{"${username_attr}", UidAttr},
+                    {"${user}", binary_to_list(Username)},
+                    {"${device_dn}", DeviceDn}],
+    PasswordString = binary_to_list(Password),
+
+    %% Here the original filter could be appended to the list of filters, like
+    %% ["and",{"objectClass", ObjectClass}]
+    %% ==> "|(&((objectClass=Class)(uiAttr=someAttr))(someKey=someValue))"
+
+    SubFilters =
+        lists:map(fun({K, V}) ->
+                          {replace_vars(K, ReplaceRules), replace_vars(V, ReplaceRules)};
+                     (Op) ->
+                          Op
+                  end, Filters),
+
+    Filter =
+        case SubFilters of
+            [] -> eldap2:equalityMatch("objectClass", ObjectClass);
+            _List -> compile_filters(SubFilters, [])
+        end,
+
+    %% auth.ldap.custom_base_dn = "${username_attr}=${user},${device_dn}"
+    BaseDN = replace_vars(CustomBaseDN, ReplaceRules),
```

Comment:
Would it worth moving these parts into their own function as it looks pretty similar to the `acl` part. (Line 73-98)